Add existing lesson

merkushin opened the pull request:
Resolves #3716

We allow to add only courseless lessons.


## Proposed Changes
* Add an option to add an existing lesson in the Course Outline block. It works only with lessons that are not attached to a course yet.

## Testing Instructions

1. Create a course.
2. In the course outline, click on Add (+) button. Choose "Existing Lesson(s)".
3. In the filter view choose lesson or a few.
4. Click Add selected.
5. It is expected to see added lessons in the updated course outline.
6. Save the course.
7. Try to look at the course on the frontend.
8. All added lessons should be there.

## Pre-Merge Checklist
<!-- Complete applicable items on this checklist **before** merging. Items that are not applicable can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed. -->
- [x] PR title and description contain sufficient detail and accurately describe the changes
- [ ] Acceptance criteria is met
- [x] Decisions are publicly documented
- [ ] Adheres to coding standards ([PHP](https://developer.wordpress.org/coding-standards/wordpress-coding-standards/php/), [JavaScript](https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/), [CSS](https://developer.wordpress.org/coding-standards/wordpress-coding-standards/css/), [HTML](https://developer.wordpress.org/coding-standards/wordpress-coding-standards/html/))
- [ ] All strings are translatable (without concatenation, handles plurals)
- [ ] Follows our naming conventions (P6rkRX-4oA-p2)
- [ ] Hooks (p6rkRX-1uS-p2) and functions are documented
- [x] New UIs are responsive and use a [mobile-first approach](https://zellwk.com/blog/how-to-write-mobile-first-css/)
- [ ] New UIs match the designs
- [ ] Different user privileges (admin, teacher, subscriber) are tested as appropriate
- [x] Code is tested on the minimum supported PHP and WordPress versions
- [ ] User interface changes have been tested on the latest versions of Chrome, Firefox and Safari
- [ ] "Needs Documentation" label is added if this change requires updates to documentation
- [ ] Known issues are created as new GitHub issues

merkushin committed:
Copy adding of existing questions from quizzes

merkushin committed:
Extract content duplicators from Sensei_Admin

merkushin committed:
Make the existing lessons modal open in the admin

merkushin committed:
wip: adding the rest endpoint

merkushin committed:
wip

merkushin committed:
working draft

merkushin committed:
add information about lesson's course

merkushin committed:
duplicate only lessons that are already assigned to a course

merkushin committed:
Extract lesson prepare method into a separate controller

merkushin committed:
Fix linter issues

merkushin committed:
Fix styles for existing lessons modal

merkushin committed:
Add tests for components

merkushin committed:
Add changelog entry

merkushin committed:
Fix duplication logic and tests

merkushin committed:
Use clock to make the test more reliable

merkushin committed:
Fix linter issues

merkushin committed:
Add generic tests for content duplicators

merkushin committed:
Fix namespace in a test

renatho commented on assets/blocks/course-outline/outline-block/existing-lessons-modal/lessons.js:
```diff
@@ -0,0 +1,209 @@
+
+	if ( ! lessons ) {
+		return (
+			<div className="sensei-lms-quiz-block__existing-lessons-modal__existing-lessons sensei-lms-quiz-block__existing-lessons-modal__existing-lessons--loading">
```
I think the prefix `sensei-lms-quiz-block` is not correct here, right? The same applies to some other places.

renatho commented on assets/blocks/course-outline/outline-block/existing-lessons-modal/lessons-modal-style.scss:
```diff
@@ -0,0 +1,105 @@
+.sensei-lms-quiz-block {
+	//
```
An empty comment.

renatho commented on assets/blocks/course-outline/outline-block/existing-lessons-modal/lessons-modal-style.scss:
```diff
@@ -0,0 +1,105 @@
+.sensei-lms-quiz-block {
```
The prefix is not correct, which makes it apply only in the cases where it's not correct in the components too.

renatho commented on assets/blocks/course-outline/outline-block/existing-lessons-modal/lessons.js:
```diff
@@ -0,0 +1,209 @@
+						{ title }
+					</label>
+				</td>
+				<td className={ courseNotFoundClass }>{ course }</td>
```
> Do you think we need to communicate the underlying logic somehow?

Maybe we could add an information icon here with a tooltip saying that it will be duplicated?

merkushin replied:
Yep, sounds good. 
Do you know if we have something like that in Sensei or a best practice? 
(I'm also looking on my own; I asked just in case you already know.)

renatho replied:
I searched for the `info` icon, and I found this: https://github.com/Automattic/sensei/pull/4001

Probably not exactly what you want here, but maybe some parts could be useful as reference?

merkushin replied:
No need for that anymore :)

renatho commented on assets/blocks/course-outline/outline-block/existing-lessons-modal/filter.js:
```diff
@@ -0,0 +1,51 @@
+ * @param {Function} props.setFilters Filters state setter.
+ */
+const Filter = ( { filters, setFilters } ) => {
+	const { searchValue } = useState( filters.search );
```
The `useState` returns an array. So the `searchValue` is always `undefined` here. With this, the `InputControl` behaves as an uncontrolled component and that's why it's working.

Given that you're getting the value on the `onChange` and you don't use the `searchValue`, maybe we could just remove this state?

renatho commented on assets/blocks/course-outline/outline-block/existing-lessons-modal/index.js:
```diff
@@ -0,0 +1,69 @@
+					) }
+				</Notice>
+			) }
+			<Filter filters={ filters } setFilters={ setFilters } />
```
I think this came from somewhere else with more filters, right? We could probably simplify it (the state and the related parts even in the other components) since we have only one search field, right?

Anyway, if you prefer to keep it, in case we add some other filters in the future, I'm okay. ;)

merkushin replied:
I removed the usage of `useState` from the Filter component, but left the component itself.
I think it still makes sense even if we won't add more filters. (But I think we might want to add a filter by a course, for example.) I mean it is tiny though it is a self-sufficient component :)

renatho commented on assets/blocks/course-outline/outline-block/existing-lessons-modal/actions.js:
```diff
@@ -0,0 +1,72 @@
+			) }
+			<li>
+				<Button
+					disabled={ isAddingSelected }
```
I think it could also be disabled if `selectedLessonIds.length === 0`. WDYT?

renatho commented on assets/blocks/course-outline/outline-block/existing-lessons-modal/index.js:
```diff
@@ -0,0 +1,69 @@
+ * @param {string}   props.clientId Outline block ID.
+ * @param {Function} props.onClose  Close callback.
+ */
+const LessonsModal = ( { clientId, onClose } ) => {
```
We could this component as `ExistingLessonsModal`, so it matches the folder where it is. WDYT?

renatho commented on assets/blocks/course-outline/outline-block/existing-lessons-modal/lessons.js:
```diff
@@ -0,0 +1,209 @@
+						return lesson;
+					}
+
+					let course = courses.find(
```
Just an optimization suggestion here: When getting the courses, we could run a `keyBy` with the `id` index, so we don't need to run `find` in the courses array for every lesson. We could just get by the key directly.

renatho commented on assets/blocks/course-outline/outline-block/existing-lessons-modal/lessons.js:
```diff
@@ -0,0 +1,209 @@
+	).map( ( block ) => block.attributes?.id );
+
+	// Lessons by current filter.
+	let lessons = useSelect(
```
Since the logic to get the `lessons` is a little more complicated and big, WDYT to extract it to a function, making it easier to understand/maintain? I think it could be here in the same file though.

merkushin replied:
Extracted here: https://github.com/Automattic/sensei/pull/7368/commits/370694388a6771da970948f88105389e17649888

renatho commented on assets/blocks/course-outline/outline-block/existing-lessons-modal/lessons.js:
```diff
@@ -0,0 +1,209 @@
+				<tbody>
+					{ lessons.length === 0 ? (
+						<tr>
+							<td colSpan="4">
```
I think the `colSpan` here should be `3` instead, right?

renatho commented on assets/blocks/course-outline/outline-block/existing-lessons-modal/use-add-existing-lessons.js:
```diff
@@ -0,0 +1,94 @@
+import apiFetch from '@wordpress/api-fetch';
+
+/**
+ * Lesson type difinition.
```
```suggestion
 * Lesson type definition.
```

renatho commented on assets/blocks/course-outline/outline-block/existing-lessons-modal/use-add-existing-lessons.js:
```diff
@@ -0,0 +1,94 @@
+ */
+
+/**
+ * Find a question block based on question ID, or title if ID is missing.
```
```suggestion
 * Find a lesson block based on lesson ID, or title if ID is missing.
```

renatho commented on assets/blocks/course-outline/outline-block/existing-lessons-modal/use-add-existing-lessons.js:
```diff
@@ -0,0 +1,94 @@
+		// Put this before the auto-block.
+		let insertIndex = nextInsertIndex;
+
+		return apiFetch( {
```
Have you explored alternatives to duplicate the posts only after publishing the lesson? I'm just thinking that when the user interacts with the editor, they don't expect something happening until they publish it.

merkushin replied:
I briefly considered an option of doing that on the saving step.
But if I understand it correctly, it requires significant changes in other part parts of the outline. Because we need to substitute IDs when we going to duplicate a lesson. In the best-case scenario, I think, it just causes double saving with a bit glitchy title (different statuses of that lesson might be shown). On the other hand, it might be not so painful. (But honestly, I don't see an issue with duplication as well. However, it might be due my bias here).
Do you think we should re-do it that way?

renatho replied:
Personally, I don't like the behavior of creating the lessons before the action of publishing because I imagine users don't expect this. If they give up the lesson duplication and later they see the lesson created, it could be confusing.

But maybe we could also go with this as it is, and change it in the future if we hear complaints.

cc @donnapep to see if she has a better opinion about this from a product perspective (Notice that the lesson post is duplicated when adding to the editor, but it's associated with the course only after saving it).

donnapep replied:
Given that the next step for this flow will be to create / use a [synced pattern](https://wordpress.org/documentation/article/reusable-blocks/) when a user selects an existing lesson, rather than duplicating lessons, what do you think about putting this PR behind a feature flag (or in a feature branch if that's easier), and only releasing once the synced pattern part is implemented? This way we don't have to do any duplication at all. And also there's no confusion for the user as to why lessons were duplicated.

merkushin replied:
Eventually, I tried to implement it the way @renatho suggested. Here is my attempt: https://github.com/Automattic/sensei/pull/7469

Unfortunately, it causes the infinite saving loop, because the content from the server comes different then the one in the editor: IDs mismatch. And it saves it again and again, duplicating lessons again and again :) Be prepared to get a lot of duplicates while testing it :)

I'd prefer to go further with the solution from that PR instead of moving the one in this PR behind the feature flag. (Just because using the feature flag here feels like a failure for me :)) However, I didn't find a good solution how to resolve the issue with matching structure. If you have no ideas either, I'm adding a feature flag.

merkushin replied:
As we show only courses that are not attached to a lesson, what do you think is better to display here? I mean how to communicate that we allow to add only "courseless" lessons?

![CleanShot 2024-02-07 at 04 45 24@2x](https://github.com/Automattic/sensei/assets/329356/d22cbfd0-e719-4128-b950-d4ceab107433)

donnapep replied:
"Courseless" isn't used much so it's a bit strange. 😄 "Available" may be a better choice.

merkushin replied:
Updated here: https://github.com/Automattic/sensei/pull/7368/commits/0b51e794cbc059e13696f31c6baf6cca2f7a0f47

![CleanShot 2024-02-08 at 14 48 09@2x](https://github.com/Automattic/sensei/assets/329356/f4b16d8e-8338-44e7-b695-913e3ec5f96b)

renatho commented on includes/rest-api/class-sensei-rest-api-lesson-actions-controller.php:
```diff
@@ -0,0 +1,198 @@
+	public function register_routes() {
+		register_rest_route(
+			$this->namespace,
+			$this->rest_base . '/prepare',
```
Regarding the name of this endpoint, and maybe the name of the class and methods too, I'm thinking that we could make it more clear with names representing the lesson duplication instead of "actions" and "preparation". WDYT?

merkushin replied:
Hum, I think it is better to leave the controller name as is, as we can add other actions here. I didn't find we create a new controller for each endpoint :)

However, I agree, I need to rename the endpoint from `prepare` to `duplicate`. "Prepare" was my previous idea that evolved into duplication :)

But, actually, I don't want to call it `duplicate` as it may duplicate it, and may not. So I want to call it `attach`.

renatho replied:
Hm, not sure if attach is a good option too. Because it's not what it's doing. It's only duplicating the posts and returning the new IDs, right? Maybe it could be a mix of the previous option and your new one? `prepare_to_attach`?

Or being more explicit? `prepare_to_attach_existing_lessons`.

Or something like `maybe_duplicate`? Or is it too bad for an endpoint? haha

merkushin replied:
Yeah, you're right, make sense... I postpone changes here until we decide what to do with duplication. There is a chance we won't need this controller at all if we decide to duplicate lessons on saving the structure.

merkushin replied:
The controller removed entirely.

renatho reviewed (commented):
Hi @merkushin! I couldn't test the styles very well because it didn't work (I added a comment about that).

Besides that, I added a few comments with some suggestions and some issues I found.

merkushin committed:
Fix classname for components and styles

merkushin committed:
Format styles

merkushin committed:
Fix table layout

merkushin committed:
Disable the Add button when no lessons selected

merkushin committed:
Rename existing lessons modal component

merkushin committed:
Optimize seek for a course

merkushin committed:
Remove unnecessary useState usage

merkushin committed:
Extract selectLessons function

merkushin committed:
Fix typo

Co-authored-by: Renatho De Carli Rosa <renatho@gmail.com>

merkushin committed:
Fix comment text

Co-authored-by: Renatho De Carli Rosa <renatho@gmail.com>

merkushin committed:
Rename endpoint

merkushin committed:
Fix test

merkushin commented:
@renatho Thanks for your review!
I addressed most of your comments.
I'll continue with the duplication icon and the tooltip on Monday.
There is also an open question here: https://github.com/Automattic/sensei/pull/7368#discussion_r1475100608

merkushin committed:
Duplicate lessons on the course structure saving (#7469)

Duplicate lessons on the course structure saving

merkushin committed:
Update title for the existing lessons modal

renatho commented on assets/blocks/course-outline/outline-block/existing-lessons-modal/lessons.js:
```diff
@@ -0,0 +1,165 @@
+		lessons.every( ( lesson ) => selectedLessons.includes( lesson.id ) );
+
+	const toggleAllHandler = ( checked ) => {
+		//const lessonIds = lessons.map( ( lesson ) => lesson.id );
```
A left comment here. :)

renatho commented on assets/blocks/course-outline/outline-block/existing-lessons-modal/lessons-modal-style.scss:
```diff
@@ -0,0 +1,102 @@
+.wp-block-sensei-lms-course-outline__existing-lessons-modal {
```
One more suggestion I noticed here. But not critical, so if you want to ignore to merge it earlier, maybe we could improve it in a future refactor.

Maybe we could leave the `wp-block-sensei-lms-course-outline__` prefix out of this class, and start a new block here. In order to follow the BEM standards, so we have a block, and element, and sometimes a modifier.

Otherwise, we start creating a big hierarchy having huge names. 🤭

merkushin replied:
Thanks, that was especially helpful, as I had to review all the classes. I had mistakes and typos there. It also fixed the other issue you pointed at, with content scrolling.
Fixed here: https://github.com/Automattic/sensei/pull/7368/commits/9e1d6ed78fafd4d2e4e94b51bfd935e3d66fa3d5

renatho commented on assets/blocks/course-outline/outline-block/existing-lessons-modal/lessons.js:
```diff
@@ -0,0 +1,165 @@
+		( lesson ) => ! addedLessonIds.includes( lesson.id )
+	);
+
+	const allChecked =
```
I noticed that for some reason it's not working. It's always unchecked.

renatho commented on assets/blocks/course-outline/outline-block/existing-lessons-modal/lessons.js:
```diff
@@ -0,0 +1,165 @@
+ * @return {Object[]} Lessons.
+ */
+const selectLessons = ( select, filters ) => {
+	const foundLessons = select( 'core' ).getEntityRecords(
```
I couldn't identify the reason, but I noticed that when searching for 1 char only, it's loading to search, but it's not filtering.

https://github.com/Automattic/sensei/assets/876340/ed289b0b-ec40-4ef6-9ed9-3c5d5dd17566

merkushin replied:
Filtering happens on the backend, that's a standard functionality, brought by WP.
I noticed that even with two letters the results are still not very good.
One option I see here is to send request only when we have 3+ letters.
At the same time, I see a UI problem here: as we need also load it when it the search term is empty. I mean it might become confusing when it works and when doesn't.
So, I'd prefer to leave it as is, in this case.

donnapep replied:
Could we use a similar implementation to how the existing questions modal works? We also need to fix the modal from jumping around. It should be a fixed height.

merkushin replied:
> Could we use a similar implementation to how the existing questions modal works?

It does. Here is a link to questions (there is no special logic to handle the term or results): https://github.com/Automattic/sensei/blob/9e1d6ed78fafd4d2e4e94b51bfd935e3d66fa3d5/assets/blocks/quiz/quiz-block/questions-modal/questions.js#L41-L48


>  We also need to fix the modal from jumping around. It should be a fixed height.

It was fixed earlier in https://github.com/Automattic/sensei/pull/7368/commits/9e1d6ed78fafd4d2e4e94b51bfd935e3d66fa3d5

renatho replied:
Interesting! I was debugging it to understand what was happening, and it searches even for the Gutenberg blocks comments.

Maybe we could check a way to filter by title only or something like this?

I also sent a question here p1707933193746959-slack-C02ECE3ML to see if it's really expected.

merkushin replied:
@renatho fixed here: https://github.com/Automattic/sensei/pull/7368/commits/46aea0702d5967aaa8fe166e5ff681581b8ae4c2

renatho replied:
Thank you for finding the solution and fixing it, @merkushin!

Reading it, I think it will impact any lesson query, no? If yes, maybe we should check somehow if it's the exact query we want?

merkushin replied:
@renatho makes sense. I added another param to the request to distinguish our request.
Fixed here: https://github.com/Automattic/sensei/pull/7368/commits/2f90a553f1b6f94b01b4846629144db43b614263

renatho commented on assets/blocks/course-outline/outline-block/existing-lessons-modal/lessons-modal-style.scss:
```diff
@@ -0,0 +1,102 @@
+.wp-block-sensei-lms-course-outline__existing-lessons-modal {
+	min-width: 566px;
+	max-width: 697px;
```
I think it would be nice to also set some height in the modal or in somewhere else to avoid the jumps while interacting, WDYT?

https://github.com/Automattic/sensei/assets/876340/ed289b0b-ec40-4ef6-9ed9-3c5d5dd17566

merkushin replied:
Fixed here: https://github.com/Automattic/sensei/commit/9e1d6ed78fafd4d2e4e94b51bfd935e3d66fa3d5

renatho commented on assets/blocks/course-outline/outline-block/existing-lessons-modal/lessons-modal-style.scss:
```diff
@@ -0,0 +1,102 @@
+
+	&__existing-lessons {
+		overflow: auto;
+		height: calc(100vh - 225px);
```
With scroll, it's not with a good behavior. I imagine it would be nice to keep the filters and actions fixed and scroll only in the list. WDYT?

https://github.com/Automattic/sensei/assets/876340/d0f020f6-2408-495b-935d-becf1ab10de9

Sorry if I missed something, but to not create a lot of orphan lessons, I added the `tr` directly through the HTML. But I think it would be the behavior of having many lessons, right?

merkushin replied:
Fixed here: https://github.com/Automattic/sensei/pull/7368/commits/9e1d6ed78fafd4d2e4e94b51bfd935e3d66fa3d5

renatho replied:
Thanks! It's much better already, but in some window heights, it's still with an unexpected scroll.

https://github.com/Automattic/sensei/assets/876340/87469a7f-d850-4bfa-bba9-a967b4c4b929

merkushin replied:
Thanks @renatho!
Fixed it here: https://github.com/Automattic/sensei/pull/7368/commits/cdc1d97b3cd59bd391ea952662a0b8ccf53c1809

renatho commented on assets/blocks/course-outline/outline-block/existing-lessons-modal/use-add-existing-lessons.test.js:
```diff
@@ -0,0 +1,67 @@
+			},
+		];
+
+		apiFetch.mockResolvedValue( lessons );
```
I think we should clean the tests now that we removed the fetch, right?

renatho commented on config/psalm/psalm-baseline.xml:
```diff
@@ -5669,6 +5654,9 @@
       <code>Sensei_REST_API_Import_Controller</code>
     </PropertyNotSetInConstructor>
   </file>
+  <file src="includes/rest-api/class-sensei-rest-api-lesson-actions-controller.php">
```
I think this file doesn't exist anymore, right?

renatho commented on config/psalm/psalm-baseline.xml:
```diff
@@ -6365,11 +6353,6 @@
       <code>(int) $offset</code>
     </RedundantCast>
   </file>
-  <file src="includes/wpml/class-sensei-wpml.php">
```
Is it here on purpose? Or did it come accidentally from another PR?

merkushin replied:
It was removed automatically by Psalm. Probably didn't have that issue anymore.

renatho commented on includes/class-sensei-course-structure.php:
```diff
@@ -5,6 +5,9 @@
  * @package sensei
  */
 
+use Sensei\Admin\Content_Duplicators\Lesson_Quiz_Duplicator;
```
I think we need a cleanup to remove all the things we're not using anymore, right?

renatho replied:
I think we need to also remove some classes and tests, right? Or are you keeping it for the future when we need it? If yes, maybe we could add it to a draft PR. WDYT?

renatho reviewed (changes requested):
Hi @merkushin!

Thank you for the changes! I'm sending a new review round.

Also, I was thinking a bit about the lesson duplication when the lesson is associated with another course. I thought without digging into the code, so there are good chances to be missing important pieces in my idea, which could invalidate my idea.

Have you thought of implementing it in the logic of the `Sensei_Course_Structure` with a new property in the lessons? Maybe something like `duplicatedFrom`. So we'd add to the course outline a lesson without ID (as a new one), but including a `duplicatedFrom`. When it's received in the server, instead of creating a new post, it would duplicate the existing one, and it would return the new ID to the frontend. Does it make sense to you?

But don't rush to implement this if it makes sense. We could iterate in the future. I'm just thinking alternatives for when we want to introduce it.

merkushin committed:
Remove deleted controller from psalm baseline

merkushin committed:
Remove unused imports

merkushin committed:
Remove mocking of unused function

merkushin committed:
Fix Select all checkbox

merkushin committed:
Fix classnames and styles

merkushin commented:
> Have you thought of implementing it in the logic of the Sensei_Course_Structure with a new property in the lessons? Maybe something like duplicatedFrom. So we'd add to the course outline a lesson without ID (as a new one), but including a duplicatedFrom. When it's received in the server, instead of creating a new post, it would duplicate the existing one, and it would return the new ID to the frontend. Does it make sense to you?

Yes, that was an only option I see to match lessons before and after the duplication. But I think it is not aligned with our plans regarding the reusable content now. 
I mean the project was supposed to be a quick one 😅 
In general, I'm ready to work on that, but not sure if it is worth doing. 

@donnapep what do you think about it?

donnapep commented:
> Yes, that was an only option I see to match lessons before and after the duplication. But I think it is not aligned with our plans regarding the reusable content now.

I think we've decided not to duplicate lessons and to only display lessons that aren't associated with any other course in the modal. We can tackle the issue of reusable lessons after this initial work ships.

merkushin committed:
Avoid scrolling for the content of the modal

merkushin committed:
In REST API search lessons by title only

merkushin committed:
Fix psalm issues

merkushin committed:
Psalm, add to baseline

merkushin committed:
Fix linter issues

merkushin committed:
Update psalm baseline

merkushin committed:
Search by title only when the request comes from add existing lesson modal

renatho reviewed (dismissed):
Thank you for the fixes, @merkushin! It looks good and works well! 🎉

I just noticed that the modal is a bit narrower than previously. Just saying in case it wasn't intentional. If this was intentional, feel free to merge it 😉

<img width="401" alt="Screenshot 2024-02-16 at 16 20 24" src="https://github.com/Automattic/sensei/assets/876340/355a8c0e-c228-46a0-b81a-ebd11b2863af">

merkushin commented:
Thank for the review @renatho!

Yes, that was intentional. On one of earlier stages I introduced a `min-width` rule. I tried to fix one of the style issues, but in fact I had to fix classnames (and you pointed me at it in one of earlier reviews).

While testing the PR on mobile, I noticed this min-width makes the modal go off the screen. I think we can do it wider for larger screens.

merkushin committed:
Make modal wider for larger screens

merkushin commented:
@renatho I decided to make it wider :) Changed here: https://github.com/Automattic/sensei/pull/7368/commits/c613a3a8e2ff3b1617713184320b56b35bf24863

![CleanShot 2024-02-16 at 13 32 42@2x](https://github.com/Automattic/sensei/assets/329356/d0005f0c-2c3d-4e1e-a5c9-b07fb9b9ee2f)
![CleanShot 2024-02-16 at 13 34 24@2x](https://github.com/Automattic/sensei/assets/329356/26aca179-cd56-4182-8497-a535c4568e55)
![CleanShot 2024-02-16 at 13 33 14@2x](https://github.com/Automattic/sensei/assets/329356/d60d6a81-88b7-4e24-85e0-52be03dbc83e)
![CleanShot 2024-02-16 at 13 33 23@2x](https://github.com/Automattic/sensei/assets/329356/b03f5eb5-f52f-432a-9c30-219f777707b8)
